From a7ff345e14ad958016c8d49bdb0d29e53c2ee772 Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Tue, 25 Apr 2017 20:50:54 +0000
Subject: [PATCH] OS-3617 libc ulwp walker should use CTF for uberdata offsets
 Reviewed By: Patrick Mooney <patrick.mooney@joyent.com> Reviewed By: Hans
 Rosenfeld <hans.rosenfeld@joyent.com> Reviewed By: Robert Mustacchi
 <rm@joyent.com> Approved By: Robert Mustacchi <rm@joyent.com>

---
 usr/src/cmd/mdb/common/modules/libc/libc.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/usr/src/cmd/mdb/common/modules/libc/libc.c b/usr/src/cmd/mdb/common/modules/libc/libc.c
index ad4c5c82ef..efd268ad55 100644
--- a/usr/src/cmd/mdb/common/modules/libc/libc.c
+++ b/usr/src/cmd/mdb/common/modules/libc/libc.c
@@ -22,7 +22,7 @@
 /*
  * Copyright (c) 2001, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 2012 by Delphix. All rights reserved.
- * Copyright 2016, Joyent, Inc.
+ * Copyright 2017, Joyent, Inc.
  */
 
 #include <sys/mdb_modapi.h>
@@ -925,11 +925,18 @@ ulwp_walk_init(mdb_walk_state_t *wsp)
 {
 	uintptr_t addr = wsp->walk_addr;
 	uintptr_t uber_addr;
+	int offset;
+
+	offset = mdb_ctf_offsetof_by_name("uberdata_t", "all_lwps");
+	if (offset == -1) {
+		offset = OFFSETOF(uberdata_t, all_lwps);
+		mdb_warn("CTF data is missing for uberdata_t; using current "
+		    "platform's offset for uberdata.all_lwps");
+	}
 
 	if (addr == NULL &&
 	    ((uber_addr = uberdata_addr()) == NULL ||
-	    mdb_vread(&addr, sizeof (addr),
-	    uber_addr + OFFSETOF(uberdata_t, all_lwps))
+	    mdb_vread(&addr, sizeof (addr), uber_addr + offset)
 	    != sizeof (addr))) {
 		mdb_warn("cannot find 'uberdata.all_lwps'");
 		return (WALK_ERR);
-- 
2.21.0

